### PR TITLE
Add cgroupv2 support in containerd/kubelet config for 22.04

### PR DESCRIPTION
**Pull request by anujmaheshwari1**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: pkg/agent/datamodel/types.go [hunk 1]
```diff
@@ -847,6 +847,14 @@ func (p *Properties) IsVHDDistroForAllNodes() bool {
 	return true
 }
 
+func (config *NodeBootstrappingConfiguration) Is2204VHD() bool {
+	if config.AgentPoolProfile.Distro == AKSUbuntuContainerd2204 || config.AgentPoolProfile.Distro == AKSUbuntuContainerd2204Gen2 {
+		return true
+	}
+
+	return false
```
Comment (alexeldeib) [inline pkg/agent/datamodel/types.go:855]:
```suggestion
        distro := config.AgentPoolProfile.Distro
        return distro == AKSUbuntuContainerd2204  || distro == AKSUbuntuContainerd2204Gen2 
```

could be nice to add a helper for e.g. IsCgroupv2Distro maybe

Comment (alexeldeib) [reply pkg/agent/datamodel/types.go:855]: (nit)

**Review by alexeldeib — APPROVED** [review]

File: pkg/agent/datamodel/types.go [hunk 2]
```diff
@@ -847,6 +847,14 @@ func (p *Properties) IsVHDDistroForAllNodes() bool {
 	return true
 }
 
+func (config *NodeBootstrappingConfiguration) Is2204VHD() bool {
+	if config.AgentPoolProfile.Distro == AKSUbuntuContainerd2204 || config.AgentPoolProfile.Distro == AKSUbuntuContainerd2204Gen2 {
```
Comment (chengliangli0918) [inline pkg/agent/datamodel/types.go:851]: Hi Anuj, could you add ' || config.AgentPoolProfile.Distro == AKSUbuntuArm64Containerd2204Gen2' to cover the ARM64 2204 distro. I added AKSUbuntuArm64Containerd2204Gen2 distro info to the Agentbaker repo last week.

File: parts/linux/cloud-init/nodecustomdata.yml [hunk 3]
```diff
@@ -432,7 +432,16 @@ write_files:
   owner: root
   content: !!binary |
     {{GetVariableProperty "cloudInitData" "containerdKubeletDropin"}}
-{{if UseRuncShimV2}}
+
+{{- if Is2204VHD}}
```
Comment (anujmaheshwari1) [inline parts/linux/cloud-init/nodecustomdata.yml:436]: @UtheMan look at this line, this file, 10-cgroupv2.conf will only be added to the custom data if "Is2204VHD" is true. this is what baker_test.go validates.

File: pkg/agent/baker_test.go [hunk 4]
```diff
@@ -563,6 +563,12 @@ var _ = Describe("Assert generated customData and cseCmd", func() {
 		Entry("AKSUbuntu1804 with IPAddress and FQDN", "AKSUbuntu1804+Containerd+IPAddress+FQDN", "1.22.2", func(config *datamodel.NodeBootstrappingConfiguration) {
 			config.ContainerService.Properties.HostedMasterProfile.FQDN = "a.hcp.eastus.azmk8s.io"
 			config.ContainerService.Properties.HostedMasterProfile.IPAddress = "1.2.3.4"
+		}),
+		Entry("AKSUbuntu2204 VHD, cgroupv2", "AKSUbuntu2204+cgroupv2", "1.24.2", func(config *datamodel.NodeBootstrappingConfiguration) {
+			config.ContainerService.Properties.AgentPoolProfiles[0].KubernetesConfig = &datamodel.KubernetesConfig{
+				ContainerRuntime: datamodel.Containerd,
+			}
+			config.ContainerService.Properties.AgentPoolProfiles[0].Distro = datamodel.AKSUbuntuContainerd2204
```
Comment (anujmaheshwari1) [inline pkg/agent/baker_test.go:571]: @UtheMan here, I specify that the distro = 2204, i.e Is2204VHD will be true

File: pkg/agent/testdata/AKSUbuntu2204+cgroupv2/CustomData [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+
+
+- path: /etc/systemd/system/kubelet.service.d/10-containerd.conf
+  permissions: "0644"
+  encoding: gzip
+  owner: root
+  content: !!binary |
+    H4sIAAAAAAAA/2yPwcrCMBCE732K0nsa/sN/EXKoGkUsFWw9iRRJFwmaTd1sir69iCgVvM0M3zDMvgYarIFDonGw5NEBssrWu6kuddPONlVTrCq9nbeLsljWSgjjkY8WgQRFZOtAETjPkIp3IAiuEQKLp/GR1d+/S38UBWDXe4usItrbREpJEeUH60YyD96cRwvmRD72QclwDwwuDxdr4It/3cqSRwAAAP//1Kx0RuEAAAA=
+- path: /etc/systemd/system/kubelet.service.d/10-cgroupv2.conf
```
Comment (anujmaheshwari1) [inline pkg/agent/testdata/AKSUbuntu2204+cgroupv2/CustomData:159]: @UtheMan and therefore, if you look at the customdata generated for 2204 test case, you ll see that it has the 10-cgroupv2.conf file but other test cases dont.

File: pkg/agent/datamodel/types.go [hunk 6]
```diff
@@ -857,6 +857,14 @@ func (p *Properties) IsVHDDistroForAllNodes() bool {
...
```
Comment (alexeldeib) [inline pkg/agent/datamodel/types.go:861]: ARM64? not resolved?

File: pkg/agent/datamodel/types.go (hunk 1 above)
Comment (anujmaheshwari1) [reply pkg/agent/datamodel/types.go]: addressed and added a helper function Is2204VHDDistro

File: pkg/agent/datamodel/types.go (hunk 2 above)
Comment (anujmaheshwari1) [reply pkg/agent/datamodel/types.go]: added a helper func that covers all 3 2204 distros

File: pkg/agent/datamodel/types.go (hunk 6 above)
Comment (anujmaheshwari1) [reply pkg/agent/datamodel/types.go]: done

**Review by alexeldeib — APPROVED** [review]

**Pull request merged**